internal/app/machined/pkg/controllers/siderolink/manager.go

```diff
@@ -110,59 +137,56 @@ func parseAPIEndpoint(sideroLinkParam string) (apiEndpoint, error) {
 
 // Run implements controller.Controller interface.
 //
-//nolint:gocyclo
+//nolint:gocyclo,cyclop
 func (ctrl *ManagerController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
-	if ctrl.Cmdline == nil || ctrl.Cmdline.Get(constants.KernelParamSideroLink).First() == nil {
-		// no SideroLink command line argument, skip controller
-		return nil
+	wgClient, wgClientErr := wgctrl.New()
+	if wgClientErr != nil {
+		return wgClientErr
 	}
 
-	s, err := smbios.GetSMBIOSInfo()
-	if err != nil {
-		return fmt.Errorf("error reading node UUID: %w", err)
-	}
-
-	nodeUUID := s.SystemInformation.UUID
-
-	var zeroKey wgtypes.Key
-
-	if bytes.Equal(ctrl.nodeKey[:], zeroKey[:]) {
-		ctrl.nodeKey, err = wgtypes.GeneratePrivateKey()
-		if err != nil {
-			return fmt.Errorf("error generating Wireguard key: %w", err)
+	defer func() {
+		if closeErr := wgClient.Close(); closeErr != nil {
+			logger.Error("failed to close wg client", zap.Error(closeErr))
 		}
-	}
-
-	apiEndpoint := *ctrl.Cmdline.Get(constants.KernelParamSideroLink).First()
-
-	parsedEndpoint, err := parseAPIEndpoint(apiEndpoint)
-	if err != nil {
-		logger.Warn("failed to parse siderolink kernel parameter", zap.Error(err))
-	}
-
-	var transportCredentials credentials.TransportCredentials
+	}()
 
-	if parsedEndpoint.Insecure {
-		transportCredentials = insecure.NewCredentials()
-	} else {
-		transportCredentials = credentials.NewTLS(&tls.Config{})
-	}
-
-	conn, err := grpc.DialContext(ctx, parsedEndpoint.Host, grpc.WithTransportCredentials(transportCredentials))
-	if err != nil {
-		return fmt.Errorf("error dialing SideroLink endpoint %q: %w", apiEndpoint, err)
-	}
-
-	sideroLinkClient := pb.NewProvisionServiceClient(conn)
+	ticker := time.NewTicker(30 * time.Second)
+	defer ticker.Stop()
 
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
+		case <-ticker.C:
```

Comment:
there's a small problem here - the reconnect logic now runs on any change to inputs, and some of that might be fine, but not all of changes are fine - e.g. `NetworkStatus` changes should not force a reconnect probably.

to address this issue, you could split up the code a bit to switch inputs: first, wait for the network to be ready, then drop it from the inputs.

example can be found here: https://github.com/siderolabs/talos/blob/339986db9d3675b78ce0d268f799ad654862fb0f/internal/app/machined/pkg/controllers/k8s/kubelet_service.go#L74-L145

Reply:
Yes, I was not very comfortable about that input. Splitting makes sense.

Reply:
controller-runtime allows to change inputs at any point in the lifetime of the controller, the only caveat is that it doesn't ever reset it back to `Inputs()` (that is only initial setting, when controller is created). 

once inputs are updated, it makes sense to `r.QueueReconcile()` to trigger at least a single run, but after that any change to inputs will trigger controller normal way.

Reply:
Updated to do exactly that.